### PR TITLE
Initial support for detailed diff of changes.

### LINKED ISSUE #11: diff_stix: detailed changes
As a user of ATT&CK and the diff_stix script, I want to be able to see granular details about what changed for each object in an update.

Add a --detailed-changes flag which reports on which fields have changed between updates (though not the contents of those changes). It should also report on whether or not relationships have been added, changed, and removed.

### PR DESCRIPTION
**Pull request by nickamon**
Initial commit of beta code for supporting detailed diff of ATT&CK changes.  I envision there will be additional refactoring required but wanted to give early access to the code in order to facilitate feedback.

Example invocation:
```
python3 diff_stix.py  -types technique -domains enterprise-attack -markdown --no-details x_mitre_platforms --no-details x_mitre_data_sources --no-details external_references --no-details x_mitre_contributors
```

@isaisabel

### CODE REVIEW

File: scripts/diff_stix.py [hunk 1]
```diff
@@ -105,6 +107,10 @@ def __init__(
         self.types = types
         self.use_taxii = use_taxii
         self.verbose = verbose
+        self.change_details = {"num_changes": 0}
```
Comment (isaisabel) [inline scripts/diff_stix.py:110]: The change quantity is better conveyed by the length of the `changes` array in the given domain and object type. Storing the count alongside the id=>change lookup is not best practice IMO.

File: scripts/diff_stix.py [hunk 2]
```diff
@@ -282,6 +288,9 @@ def load_taxii(new=False):
                         if new_version > old_version:
                             # an update has occurred to this object
                             changes.add(key)
+                            result = diff(old["id_to_obj"][key], new["id_to_obj"][key])
+                            attack_id = new["id_to_obj"][key]['external_references'][0]['external_id']
+                            self.change_details[attack_id] = result
```
Comment (isaisabel) [inline scripts/diff_stix.py:293]: ID used as the key should be the STIX ID in case of ATT&CK ID conflict. There shouldn't be ATT&CK ID conflicts but it's always best to use STIX ID in these cases.

Comment (isaisabel) [reply scripts/diff_stix.py:293]: `minor_changes` also needs change detection. See line 299 below.

File: scripts/diff_stix.py [hunk 3]
```diff
@@ -420,19 +429,70 @@ def placard(item):
                     return f"[{item['name']}]({self.site_prefix}/{self.getUrlFromStix(item, is_subtechnique)})"
 
 
+            def change_details_placard(item, tab_count = 2):
```
Comment (isaisabel) [inline scripts/diff_stix.py:432]: This method needs a docstring comment to describe the parameters, return type, and general usage.

File: scripts/diff_stix.py [hunk 4]
```diff
@@ -420,19 +429,70 @@ def placard(item):
...
                     if grouping["parentInSection"]:
                         sectionString += f"\t* {placard(child) }\n"
+                        if section == "changes":
+                            sectionString += f"{change_details_placard(child) }\n"
                     else:
                         sectionString += f"* { grouping['parent']['name'] }: { placard(child) }\n"
+                        if section == "changes":
+                            sectionString += f"{change_details_placard(child, 1) }\n"
 
+            print(f"Number of (sub-)technique changes in section {section} is {self.change_details['num_changes']}")
```
Comment (isaisabel) [inline scripts/diff_stix.py:495]: Changes applies to more than just techniques and sub-techniques, so this count line should be more generic.

File: scripts/diff_stix.py [hunk 5]
```diff
@@ -651,6 +711,12 @@ def layers_dict_to_files(outfiles, layers):
         action="store_true",
         help="Add a key explaining the change types to the markdown"
     )
+
+    parser.add_argument("--no-details",
+        type=str,
+        action="append",
+        help="Skip providing details for the specified field"
+    )
```
Comment (isaisabel) [inline scripts/diff_stix.py:719]: The way argparse is usually used in ATT&CK and `attack-scripts` is that we use `nargs="+"` instead of `action="append"` for multi-value arguments. That should be done here as well.

Comment (isaisabel) [reply scripts/diff_stix.py:719]: The presence of details in the first place should be opt-in, like `--minor-changes` is already. So we need another argument to turn on detailed changes mode.

File: scripts/diff_stix.py (hunk 4 above)
Comment (isaisabel) [reply scripts/diff_stix.py:495]: This also makes more sense as line at the end of the output change section (`sectionString`), since seeing it in the output print isn't that useful. It also interrupts the "generating markdown string... done" output when `--verbose` is specified:

```
generating markdown string... Number of (sub-)technique changes in section additions is 0
Number of (sub-)technique changes in section additions is 0
Number of (sub-)technique changes in section minor_changes is 0
Number of (sub-)technique changes in section deletions is 0
done
```

File: scripts/diff_stix.py (hunk 5 above)
Comment (isaisabel) [reply scripts/diff_stix.py:719]: We may also want multiple modes of details. Something like this perhaps:
```
parser.add_argument("--details",
    choices=["fields", "values"]
    help="provide detailed report of what changed for each object. If 'fields' is specified, it will report on what fields changed. If 'values' is specified, it will report on how the value of each field changed as well."
)
```
This will help avoid excessively long outputs for users who just want a bullet-point ("changed: description, data_sources") list for each object instead of the full before/after for each field which might not be nessecary for every use case.

File: scripts/diff_stix.py [hunk 6]
```diff
@@ -420,19 +429,70 @@ def placard(item):
...
+                changes = []
+                for change in details:
+                    what_changed, index_changed = (f"{change[1][0]}", change[1][1]) if type(change[1]) == list else (change[1], None)
+                    if what_changed not in self.no_details:
+                        self.change_details["num_changes"] += 1
+                        how_changed = change[2]
+                        change_type = change[0]
+                        indent = "\t" * tab_count
+                        if change_type == "add":
+                            changes.append(f"{indent}* Added entries to {what_changed} field")
```
Comment (isaisabel) [inline scripts/diff_stix.py:444]: This doesn't seem to always work -- what_changed is empty in some cases leading to a double space in the output:

```
	* Added entries to  field
		* x_mitre_is_subtechnique

		```
		False
		```

	* Removed entries from  field
		* x_mitre_contributors:  ['Itamar Mizrahi', 'Tristan Bennett, Seamless Intelligence']
```
I'm not sure exactly what the cause is.

**Review by isaisabel — CHANGES_REQUESTED** [review]
Overall this looks excellent. There are a few improvements to be made to be flexible for other use cases which I've documented below

Two other improvements:
- currently the output for descriptions and detection notes is hard to parse through since those fields can get quite long. For `string` value fields it may be worth changing so that it does an inline diff of the field instead of a before/after diff. That would help catch exactly which part of the description changed. See [this stackoverflow post](https://stackoverflow.com/a/788780) for an example implementation creating an inline diff of two strings.
- In cases where there is a difference detected but no fields detected (because they were excluded), there needs to be a marker explaining this. For example, there are cases where the version number increments but no other fields change because ATT&CK added/updated  relationships with the object. We should add a marker saying something like `- (no field changes detected)` in those cases to show that there isn't a bug omitting changes being listed there and to prevent the blank line for changes which currently appears. Here's what that sort of output (no changes) looks like currently:
```
* [Software Configuration](/mitigations/M1054)

* [User Account Control](/mitigations/M1052)

* [User Account Management](/mitigations/M1018)

* [User Training](/mitigations/M1017)
	* Changed description field
		* Old:  
		```
		Train users to to be aware of access or manipulation attempts by an adversary to reduce the risk of suc[...]
```

**Comment by jondricek** [comment]
Admin note: closing all remaining issues and pull requests prior to archiving the repository
